### PR TITLE
polkadot: 0.2.17 -> 0.8.24

### PR DESCRIPTION
**Pull request by andresilva**
###### Things done

- [x] Tested using sandboxing ([nix.useSandbox](https://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](https://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [ ] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [ ] Ensured that relevant documentation is up to date
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

### CODE REVIEW

File: pkgs/applications/blockchains/polkadot/default.nix [hunk 1]
```diff
@@ -1,37 +1,52 @@
...
+let
+  mozillaOverlay = fetchFromGitHub {
+    owner = "mozilla";
+    repo = "nixpkgs-mozilla";
+    rev = "efda5b357451dbb0431f983cca679ae3cd9b9829";
+    sha256 = "11wqrg86g3qva67vnk81ynvqyfj0zxk83cbrf0p9hsvxiwxs8469";
+  };
+  mozilla = callPackage "${mozillaOverlay.out}/package-set.nix" { };
+  rustWasm = (mozilla.rustChannelOf { date = "2020-09-17"; channel = "nightly"; }).rust.override {
+    targets = [ "wasm32-unknown-unknown" ];
```
Comment (andresilva) [inline pkgs/applications/blockchains/polkadot/default.nix:18]: I don't know how to go about doing this without using the mozilla overlay. We need to have a rustc that supports both the host target and also wasm32. The builtin nix rust support seems to only support defining one target environment.

**Review by bkchr — COMMENTED** [review]
Just build with `SKIP_WASM_BUILD=1` and than you don't need nightly.

File: maintainers/maintainer-list.nix [hunk 2]
```diff
@@ -520,6 +520,12 @@
     githubId = 293191;
     name = "Andres Loeh";
   };
+  andresilva = {
+    email = "andre.beat@gmail.com";
+    github = "andresilva";
+    githubId = 123550;
+    name = "André Silva";
+  };
```
Comment (jonringer) [inline maintainers/maintainer-list.nix:528]: Please make this into a separate commit:

```
maintainers: add <name>
```

**Comment by jonringer** [comment]
@GrahamcOfBorg eval

File: maintainers/maintainer-list.nix (hunk 2 above)
Comment (andresilva) [reply maintainers/maintainer-list.nix:528]: Done.

**Review by akru — APPROVED** [review]
I've built it locally, works well

File: pkgs/applications/blockchains/polkadot/default.nix [hunk 3]
```diff
@@ -1,37 +1,43 @@
-{ stdenv
+{ clang
 , fetchFromGitHub
+, llvmPackages
+, protobuf
 , rustPlatform
-, pkgconfig
-, openssl
+, stdenv
```
Comment (jonringer) [inline pkgs/applications/blockchains/polkadot/default.nix:6]: preference to eliminate unnecessary stdenv usage
```suggestion
, lib
```

File: pkgs/applications/blockchains/polkadot/default.nix [hunk 4]
```diff
@@ -1,37 +1,43 @@
...
+  # NOTE: We don't build the WASM runtimes since this would require a more
+  # complicated rust environment setup. The resulting binary is still useful for
+  # live networks since those just use the WASM blob from the network chainspec.
+  BUILD_DUMMY_WASM_BINARY = 1;
 
-  buildInputs = [ pkgconfig openssl openssl.dev ];
+  # We can't run the test suite since we didn't compile the WASM runtimes.
+  doCheck = false;
 
   meta = with stdenv.lib; {
```
Comment (jonringer) [inline pkgs/applications/blockchains/polkadot/default.nix:36]:
```suggestion
  meta = with lib; {
```

File: pkgs/applications/blockchains/polkadot/default.nix [hunk 5]
```diff
@@ -1,37 +1,43 @@
...
     description = "Polkadot Node Implementation";
     homepage = "https://polkadot.network";
     license = licenses.gpl3;
-    maintainers = [ maintainers.akru ];
+    maintainers = [ maintainers.akru maintainers.andresilva ];
```
Comment (jonringer) [inline pkgs/applications/blockchains/polkadot/default.nix:40]:
```suggestion
    maintainers = with maintainers; [ akru andresilva ];
```

File: pkgs/applications/blockchains/polkadot/default.nix (hunk 3 above)
Comment (andresilva) [reply pkgs/applications/blockchains/polkadot/default.nix]: Updated and rebased.

File: pkgs/applications/blockchains/polkadot/substrate-wasm-builder-runner.patch [hunk 6]
```diff
@@ -0,0 +1,25 @@
...
+index 78047a1a..2d571f8e 100644
+--- a/Cargo.toml
++++ b/Cargo.toml
+@@ -112,3 +112,6 @@ polkadot = { path = "/usr/bin/polkadot" }
+ 
+ [package.metadata.rpm.files]
+ "../scripts/packaging/polkadot.service" = { path = "/usr/lib/systemd/system/polkadot.service", mode = "644" }
++
++[patch.crates-io]
++substrate-wasm-builder-runner = { git = "https://github.com/paritytech/substrate", branch = "master" }
```
Comment (jonringer) [inline pkgs/applications/blockchains/polkadot/substrate-wasm-builder-runner.patch:25]: just noticed that this follows master, but that's not a stable endpoint... an update to their master will break the build. Please use a commit instead

Comment (andresilva) [reply pkgs/applications/blockchains/polkadot/substrate-wasm-builder-runner.patch:25]: https://github.com/NixOS/nixpkgs/pull/98785/files#diff-71369739716eac7dd99e9da07e1f4944R11

That's done here in the `Cargo.lock`.

**Comment by RaghavSood** [comment]
Just adding in that I've synced a full, archival node to be in sync with the network - works fine

**Review by akru — APPROVED** [review]

**Comment by andresilva** [comment]
ping

**Comment by mmahut** [comment]
@GrahamcOfBorg build polkadot

**Comment by FlorianFranzen** [comment]
@andresilva Very nice for getting this going, I am maintaining a similar derivation [here](https://gitlab.w3f.tech/florian/w3fpkgs/-/blob/master/pkgs/parity-polkadot/default.nix) which also builds the wasm runtime. I will give this a review Monday.

**Comment by andresilva** [comment]
@FlorianFranzen the initial version of this PR was also building the runtime, but this required pulling nightly rustc with wasm environment from Mozilla overlay, and I'm not sure if this would be accepted here. In the end anyone that cares about building the runtime should probably be building from source, this is mostly for regular users that just want to sync the chain.

**Review by jonringer — COMMENTED** [review]
if https://github.com/paritytech/substrate ever gets a bump to master, this build will fail (cargoDeps hash will change). Please use an immutable source like a commit or tag

**Comment by andresilva** [comment]
Substrate has had multiple bumps to master since I have opened this PR. The way cargo works is that you define some source for a dependency in the `Cargo.toml` file and the specific source is fixed in the `Cargo.lock` file. So even though in the `Cargo.toml` we are referencing `master`, the `Cargo.lock` fixes a specific commit so that everyone builds the same thing every time.

You can have a look at how dependencies are defined in polkadot for example (https://github.com/paritytech/polkadot/blob/master/service/Cargo.toml#L21) where substrate master is also referenced, but the specific commit that they point to is defined in the [`Cargo.lock`](https://github.com/paritytech/polkadot/blob/master/Cargo.lock).

As far as this PR goes I had to patch one cargo dependency in order to use a more recent version, I did this by pointing a dependency to the git master branch version https://github.com/NixOS/nixpkgs/blob/5bafcbf7c184e93b6898c5faf5b2f8bde8455027/pkgs/applications/blockchains/polkadot/substrate-wasm-builder-runner.patch#L25 in the `Cargo.toml` file, but the exact commit that will be checked out is defined here https://github.com/NixOS/nixpkgs/blob/5bafcbf7c184e93b6898c5faf5b2f8bde8455027/pkgs/applications/blockchains/polkadot/substrate-wasm-builder-runner.patch#L11 in the `Cargo.lock` file.

**Comment by jonringer** [comment]
oh yea, forgot about the lock file nullifying the toml

**Comment by andresilva** [comment]
ping

**Comment by mmahut** [comment]
This looks good, thank you.
